File: src/main/py/logs_analysis.py
from pyspark.sql import SparkSession
from pyspark.sql.types import Row

# Create the spark session

spark = SparkSession.builder.master("local")\
    .appName("count-words-python")\
    .getOrCreate()
   
ip_database = {}
with open("/Users/shad/ip.db") as f:
    table = f.read().split("\n")
    for row in table:
        attr = row.split(",")
        if len(attr) > 1:
            ip_database[attr[0]] = attr[1]
        

# print(get_country_http("122.180.169.236"))

website_logs = spark.read.json("/Users/shad/logs.json")
with_country = website_logs.rdd.map(lambda row: Row(ip_addr = row.asDict()["ip_addr"], user_id = row.asDict()["user_id"], country=ip_database[row.asDict()["ip_addr"]])).toDF()
total_users = with_country.count()
stats = with_country.groupBy("country").count().rdd.map(lambda row: Row(country = row[0], users = row[1], percentage_users = (row[1]/total_users)*100)).toDF()
stats.show()

spark.stop()

File: src/main/scala/com/gigahex/samples/analytics/LogsAnalysis.scala
package com.gigahex.samples.analytics

import java.io.{BufferedReader, InputStreamReader}
import java.net.{HttpURLConnection, URL}

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.commons.logging.LogFactory
import org.apache.spark.sql.SparkSession


object LogsAnalysis {

  val logger = LogFactory.getLog(getClass)

  def getCountry(ip: String): String = {

    val url = s"https://tools.keycdn.com/geo.json?host=${ip}"
    val USER_AGENT = "keycdn-tools:https://www.example.com"

    val obj = new URL(url)
    val con = obj.openConnection.asInstanceOf[HttpURLConnection]
    con.setConnectTimeout(1000)
    con.setRequestProperty("User-Agent", USER_AGENT)

    val in = new BufferedReader(new InputStreamReader(con.getInputStream))
    var inputLine: String = null
    val response = new StringBuffer
    inputLine = in.readLine()
    while(inputLine != null){
      response.append(inputLine)
      inputLine = in.readLine()
    }

    in.close()
    val mapper = new ObjectMapper()
    mapper.readTree(response.toString).path("data").path("geo").path("country_name").toString

  }

  def main(args: Array[String]): Unit = {
    /**
     * Setup the spark session
     */
    val spark = SparkSession.builder()
      .appName("count-words-scala")
      .getOrCreate()
    import spark.implicits._

    val websiteLogs = spark.read.json("/Users/gigahex/logs.json")
    val withCountry = websiteLogs.map(row => (
      row.getAs[String]("user_id"),
      row.getAs[String]("ip_addr"),
      getCountry(row.getAs[String]("ip_addr")))).toDF("user_id","ip", "country")
    val total = withCountry.count()
    val stats = withCountry.groupBy("country")
      .count()
      .map(row => (row.getAs[String]("country"),
        row.getAs[Long]("count"),
        (row.getAs[Long]("count")/total.toDouble)*100))
      .toDF("Country","Users", "% Users")

    stats.show()

    spark.stop()
  }

}
